README.md
# Quick Item Scroll

Quick Item Scroll is a mod that enables you to quickly change between related items.  For instance, if you're currently
holding a transport belt, the "Next Item Group" keybind will pick up a fast transport belt, and the "Next Related Item"
keybind will pick up a splitter. A wide variety of items and groups are supported, including transport belts, inserters,
train vehicles, railway and rail signals, inserters, pipes, and modules.

Item cycling can set your cursor to a ghost of items you don't currently have on you or, in cheat mode, to create items
out of thin air as you scroll. This behaviour is toggleable. You can also, on a per-player basis, set individual items
to always be included or skipped in your scroll.

Quick Item Scroll is a fork and replacement for [Quick Item Swap](https://mods.factorio.com/mod/QuickItemSwap), which
has not been updated to the latest version of Factorio.

The following mods are supported:
 
  * [Batteries Not Included](https://mods.factorio.com/mods/dewiniaid/BatteriesNotIncluded)
  * [Electric Vehicles](https://mods.factorio.com/mod/ElectricVehicles3) (locomotive only)
  * [FARL](https://mods.factorio.com/mod/FARL) (locomotive only)
  * [Creative Mode](https://mods.factorio.com/mod/creative-mode-fix)
  * [Logistic Train Network](https://mods.factorio.com/mod/LogisticTrainNetwork)
  * [Train Supply Manager](https://mods.factorio.com/mod/train-pubsub)
  * [Smarter Trains](https://mods.factorio.com/mod/SmartTrains)
  * [Vehicle Wagon](https://mods.factorio.com/mod/Vehicle%20Wagon)
  * [Nixie Tubes](https://mods.factorio.com/mod/nixie-tubes) (Just justarandomgeek's version currently)
  * [Bob's Logistics mod](https://mods.factorio.com/mod/boblogistics) [sic] (Only inserters, currently)

Additional mods that do not have built-in support can add support on their own through the remote function call
interface. Unfortunately, this support does not automatically carry over from QuickItemSwap.
   
## Known Issues

* Mousewheel bindings may not work consistently due to what appears to be a
  [bug in Factorio](https://forums.factorio.com/viewtopic.php?f=34&t=54327). It's unclear to 
 
* Not tested in multiplayer.  Please report back!

## Unknown Issues
   
Found a bug?  Please visit the [Issues page](https://github.com/alercah/quick-item-scroll/issues) to see if it has 
already been reported, and report it if not.

## Changelog

See [changelog.txt].

## API for Mod Developers 

General recommended flow is:

 - Make your mod optionally depend on quick-item-scroll so that quick-item-scroll initializes first.
 - During your mod's `on_init`, `on_configuration_changed` and possibly `on_runtime_mod_settings_changed` events, call
   `apply_patch` with the appropriate information.
 - Listen for the custom events `on_qis_mappings_reset` and, if desired, `on_qis_mappings_patched` (event IDs are 
   published through `get_events`) and resubmit your patches as needed. 

Currently implemented:

#### `remote.call("quick-item-scroll", "debug", [new_setting])`

Returns (and possibly changes) the debug setting for QIS. Note that debug being on will produce an incessant amount
of screen spam.  Set to `true` to enable debugging, `false` to disable, or `nil` to just see what the current setting is.
 
#### `remote.call("quick-item-scroll", "get_events")`

Returns a table of `string = number` containing the custom events used by quick-item-scroll.  Right now, the two events
are:
  
  - `on_quick_item_scroll_mappings_reset` - Triggered when quick-item-scroll or
    another mod resets the item mappings and rebuilds them from scratch.  This
    can happen on version updates, mod changes, etc.  When this fires, any
    patches provided by your mod will be lost and thus must be resubmitted.
  
    Event table fields:
    
    | Field | Explanation |
    | ----- | ----------- |
    | `why` | Reason for the reset.  May be `"init"` (QIS's `on_init()`), `"settings-changed"` (QIS's mod settings changing), `"configuration-changed"` (Mod versions or startup settings changing) or `"remote"` (a remote caller asked it to reset)  

  - `on_quick_item_scroll_mappings_patched` - Triggered in response to someone
    (possibly you) calling `apply_patch`.  Allows for mods to react to other
    mods patching the table if it makes sense to.   
  
    Event table fields:
    
    | Field | Explanation |
    | ----- | ----------- |
    | `patch` | Contents of the patch that was provided.
    | `source` | Source providing the patch.  This is dependant on the caller providing a meaningful `source` parameter.
    
#### `remote.call("quick-item-scroll", "get_mappings", [category, [group, [item]]])`

Returns information on currently defined mappings.  If all 3 parameters are nil, the entire table is returned.  
Otherwise, returns just a subset for a given category, group within a category, or item within a group.

#### `remote.call("quick-item-scroll", "refresh", [only_if_dirty])`

Refreshes the internal state that QIS maintains to properly cycle between things.  If the optional parameter is `true`,
the refresh is only performed if the data is marked as out of date.

This is generally unneccessary -- `apply_patch` marks the data as outdated and it will be rebuilt on the first access
that requires it when outdated.

If `player` is specified, results will be printed to that player; otherwise they are printed using `game.print`
 
Produces no output if everything is valid.

#### `remote.call("quick-item-scroll", "validate_mappings", [player])`

Prints to the console any items in the mapping table that don't have a corresponding item prototype.

If `player` is specified, results will be printed to that player; otherwise they are printed using `game.print`
 
Produces no output if everything is valid.

#### `remote.call("quick-item-scroll", "apply_patch", patch, source)`

Applies a patch to the mapping table.

`source` is the name of your mod, and is used to explains where the patch came from.  It appears in error messages and 
in the `on_qis_mappings_patched` event.  It is recommended you use the `info.json` name of your mod here for
consistency, though nothing enforces this.

`patch` consists of a table formatted similar to the mapping table.  (See `mappings/base.lua` for a reference.)  It will
be merged into the mapping table using the following rules:

- If a category does not exist, it will be created.  If it exists, the groups will be merged.  If the patch explicitly 
  sets the category to `false`, the category will be deleted.  If the patch omits the category or sets it to `nil`, it 
  will be left as-is.
  
- The above logic also applies to groups within a category (with items being merged) and items within a group.

- Groups and items may specify a `default_order` attribute which is equivalent to `order` but won't override the 
  existing value if there is one.
  
Example: You've decided yellow belts are too slow, so your mod rips them out of the game and adds ludicrous transport 
belts instead.
 
A patch accomplishing this will be look like this:

```lua
-- Note that this outer table is always required!
categories = {
    belts = {
        groups = {
            -- Deletes the 'normal' group
            -- normal = false,
                
            -- Addds our new group
            ludicrous = {
                order = 400,
                items = {
                    ['ludicrous-belt'] =             { order = 100, type = 'belt' },
                    ['ludicrous-underground-belt'] = { order = 200, type = 'underground' },
                    ['ludicrous-splitter'] =         { order = 300, type = 'splitter' },
                }
            }
        }
    }
}
```

Note that it is harmless to have items in the mapper that don't have prototypes -- though it will trigger a (probably
unnoticeable) performance hit when cycling items.

#### `remote.call("quick-item-scroll", "dump_inventory", inventory)`

Intended for development use only and may change at any time.

#### `remote.call("quick-item-scroll", "dump_quickbar", player)`

Intended for development use only and may change at any time.

#### `remote.call("quick-item-scroll", "dump_mappings")`

Writes the current mapping table to `script-output/quick-item-scroll-mappings.txt`. Intended for development use only and
may change at any time.

#### `remote.call("quick-item-scroll", "dump_overrides", player)`

Intended for development use only and may change at any time.


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.4.0
Date: 07. 20. 2020

  Changes:
    - alercah forks the mod
    - Add support for inserters in Bob's Logistics
    - Renamed "whitelist" and "blacklist" to "skip" and "include" overrides
    - Skip overridden items even when some are held in hand

---------------------------------------------------------------------------------------------------
Version: 0.3.4
Date: 06. 14. 2020

  Changes:
    - Fix crash for 'Unknown consuming type' when loading mod

---------------------------------------------------------------------------------------------------
Version: 0.3.3
Date: 01. 31. 2020

  Changes:
    - Update for Factorio 0.18

---------------------------------------------------------------------------------------------------
Version: 0.3.2
Date: 14. 03. 2019

  Changes:
    - QuickItemSwap can now cycle between item ghosts if you lack the actual item.
      - By default, this excludes items that have not been researched.  You can globally disable this in Mod Settings, or you can whitelist specific items.
    - You can now whitelist or blacklist items (default keybind: Alt+Y while holding an item)
      - When cycling between related items, blacklisted items will be ignored (unless they are in your inventory)
      - When cycling between related items, whitelisted items will be allowed (even if they are not yet researched)
      - Blacklisting and whitelisting only affects cases when a ghost would be chosen or if an item would be created (in cheat mode).  Items in your inventory are always considered.
      - You can reset your entire blacklist with /qis-clear-blacklist
   - Added support for the train stops in the Train Supply Manager mod

---------------------------------------------------------------------------------------------------
Version: 0.3.1
Date: 27. 02. 2019
  Changes:
    - Update for Factorio 0.17.
    - Removed all quickbar logic.
    - Generate group information for belts/splitters/undergrounds procedurally.  This should automatically add support for modded belts, such as those from Bob's Logistics.

---------------------------------------------------------------------------------------------------
Version: 0.3.0
Date: 23. 04. 2018
  Changes:
    - Add support for blueprints, blueprint books, deconstruction planners, and similar items from mods.
      - Items of a particular type are treated as members of one group.  "Y" will cycle between different blueprint books, for instance.
      - Cycling groups will cycle between selecting a blueprint, a blueprint book, or a deconstruction planner.
      - Other items that have an item_number are also supported when cycling items, but do not support cycling groups.
    - Added support for the Artillery Wagon
    - Add support for Electric Vehicles (locomotives only)
    - Add support for FARL (locomotives only)

---------------------------------------------------------------------------------------------------
Version: 0.2.4
Date: 23. 12. 2017
  Changes:
    - Add support for the unofficial Creative Mode Fix.
    - Add Buffer Chests to the Logistic Chests group, and reordered the group to match the inventory window.

---------------------------------------------------------------------------------------------------
Version: 0.2.3
Date: 13. 12. 2017
  Changes:
    - Update for Factorio 0.16
    - Temporarily add support for Batteries Not Included, since [a bug in the current Factorio experimental](https://forums.factorio.com/viewtopic.php?f=182&t=54567&p=321491) prevents that support from working in the opposite direction.
  Modding:
    - Added more verbose error messaging if another mod submits an invalid patch.
    - Swapped the order of arguments on `apply_patch` to make it less likely for someone to inadvertently forget to specify `source`.

---------------------------------------------------------------------------------------------------
Version: 0.2.2
Date: 11. 12. 2017
  Changes:
    - Add support for items from Logistic Train Network, Smarter Trains, Vehicle Wagon, and Nixie Tubes.
    - Add speculative support for artillery wagons.

---------------------------------------------------------------------------------------------------
Version: 0.2.1
Date: 05. 12. 2017
  Changes:
    - Fix default keybindings for Previous Item and Previous Group.
    - Console announcements when a configuration change is detected are now only displayed in debug mode.  They also now are prefixed with `[QuickItemSwap]` to indicate they are coming from this mod.

---------------------------------------------------------------------------------------------------
Version: 0.2.0
Date: 01. 12. 2017
  Changes:
    - Added support for items from Creative Mode
    - Added support for creating items from thin air (rather than skipping over them) when your player is in cheat mode (i.e. as a result of the creative mode mode or `/c game.players[1].cheat_mode = true`).  Can be toggled in settings.  Has no effect when not in cheat mode.
  Modding:
    - Made significant internal changes to allow mod integration
    - First pass of an API to allow other mods to add recipes.

---------------------------------------------------------------------------------------------------
Version: 0.1.1
Date: 30. 11. 2017
  Changes:
    - Flip logic on modules to make more sense: each tier is now a group, rather than each module type being a group.

---------------------------------------------------------------------------------------------------
Version: 0.1.0
Date: 30. 11. 2017
  Changes:
    - First public release.



data.lua
-- Add keybindings

data:extend({
    {
        type = "custom-input",
        name = "quick-item-scroll-item-prev",
        key_sequence = "CONTROL + Y",
        consuming = "game-only"
    },
    {
        type = "custom-input",
        name = "quick-item-scroll-item-next",
        key_sequence = "Y",
        consuming = "game-only"
    },
    {
        type = "custom-input",
        name = "quick-item-scroll-group-prev",
        key_sequence = "CONTROL + SHIFT + Y",
        consuming = "game-only"
    },
    {
        type = "custom-input",
        name = "quick-item-scroll-group-next",
        key_sequence = "SHIFT + Y",
        consuming = "game-only"
    },
    {
        type = "custom-input",
        name = "quick-item-scroll-toggle-override",
        key_sequence = "ALT + Y",
        consuming = "game-only"
    },
})


settings.lua
-- Cheat Mode options
data:extend({
    {
        type = "bool-setting",
        name = "quick-item-scroll-support-cheat-mode",
        default_value = true,
        setting_type = "runtime-per-user",
        order = 100,
    },
    {
        type = "bool-setting",
        name = "quick-item-scroll-use-ghosts",
        default_value = true,
        setting_type = "runtime-per-user",
        order = 220,
    },
    {
        type = "bool-setting",
        name = "quick-item-scroll-requires-tech",
        default_value = true,
        setting_type = "runtime-per-user",
        order = 221,
    },
    --    {
    --        type = "bool-setting",
    --        name = "quick-item-scroll-destroy-cheated-items",
    --        default_value = true,
    --        setting_type = "runtime-per-user",
    --        order = 110,
    --    },
})

-- Creative Mode support
if mods["creative-mode"] or mods["creative-mode-fix"] then
    data:extend({
        {
            type = "bool-setting",
            name = "quick-item-scroll-support-creative-mode",
            default_value = true,
            setting_type = "runtime-global",
            order = 300,
        },
        {
            type = "bool-setting",
            name = "quick-item-scroll-creative-mode-split",
            default_value = false,
            setting_type = "runtime-global",
            order = 310,
        },
        {
            type = "bool-setting",
            name = "quick-item-scroll-support-cheat-mode-with-creative",
            default_value = false,
            setting_type = "runtime-per-user",
            order = 320,
        },
        --{
        --    type = "string-setting",
        --    name = "quick-item-scroll-initial-override",
        --    default_value = "burner-inserter,burner-mining-drill",
        --    setting_type = "runtime-per-user",
        --    order = 220,
        --},
    })
end


mappings/boblogistics.lua
-- Not finished, and not in this release.  Bob's just adds more stuff than I feel like supporting at the moment.
-- (Maybe you can ask him to add support on his end instead?)

return function()
    local result = {}

    if not game.active_mods['boblogistics'] then
        return nil
    end

    return {
        categories = {
          --[[
            trains = {
                groups = {
                    trains = {
                        default_order = 100,
                        items = {
                            ["bob-locomotive-2"] = { order = 102 },
                            ["bob-locomotive-3"] = { order = 103 },
                            ["bob-armoured-locomotive"] = { order = 110 },
                            ["bob-armoured-locomotive-2"] = { order = 111 },
                            ["bob-cargo-wagon-2"] = { order = 202 },
                            ["bob-cargo-wagon-3"] = { order = 203 },
                            ["bob-armoured-cargo-wagon"] = { order = 210 },
                            ["bob-armoured-cargo-wagon-2"] = { order = 211 },
                            ["bob-fluid-wagon-2"] = { order = 302 },
                            ["bob-fluid-wagon-3"] = { order = 303 },
                            ["bob-armoured-fluid-wagon"] = { order = 310 },
                            ["bob-armoured-fluid-wagon-2"] = { order = 311 },
                        }
                    }
                }
            },
          ]]--
            inserters = {
                groups = {
                    plain = {
                        order = 200,
                        items = {
                            ['yellow-filter-inserter'] = { order = 300, type = 'filter' },
                            ['long-handed-inserter'] = false,
                        },
                    },
                    fast = {
                        order = 300,
                        items = {
                            ['fast-inserter'] = false,
                            ['filter-inserter'] = false,
                            ['long-handed-inserter'] = { order = 100, type = 'normal' },
                            ['red-filter-inserter'] = { order = 200, type = 'filter' },
                            ['red-stack-inserter'] = { order = 300, type = 'stack' },
                            ['red-stack-filter-inserter'] = { order = 400, type = 'stack-filter' },
                        },
                    },
                    stack = false,
                    express = {
                        order = 400,
                        items = {
                            ['fast-inserter'] = { order = 100, type = 'normal' },
                            ['filter-inserter'] = { order = 200, type = 'filter' },
                            ['stack-inserter'] = { order = 300, type = 'stack' },
                            ['stack-filter-inserter'] = { order = 400, type = 'stack-filter' },
                        },
                    },
                    turbo = {
                        order = 500,
                        items = {
                            ['turbo-inserter'] = { order = 100, type = 'normal' },
                            ['turbo-filter-inserter'] = { order = 200, type = 'filter' },
                            ['turbo-stack-inserter'] = { order = 300, type = 'stack' },
                            ['turbo-stack-filter-inserter'] = { order = 400, type = 'stack-filter' },
                        },
                    },
                    ultimate = {
                        order = 600,
                        items = {
                            ['express-inserter'] = { order = 100, type = 'normal' },
                            ['express-filter-inserter'] = { order = 200, type = 'filter' },
                            ['express-stack-inserter'] = { order = 300, type = 'stack' },
                            ['express-stack-filter-inserter'] = { order = 400, type = 'stack-filter' },
                        },
                    },
                }
            },
        },
        --[[
        fluids = {
            groups = {
                pipes = {
                    default_order = 100,
                    items = {
                        ['bob-valve'] = { order = 1000 },
                    }
                },
                structures = {
                    default_order = 200,
                    items = {
                        ['offshore-pump'] = { order = 100 },
                        ['bob-pump-2'] = { order = 202 },
                        ['bob-pump-3'] = { order = 203 },
                        ['bob-pump-4'] = { order = 204 },
                        ['storage-tank'] = { order = 300 },
                        ['storage-tank-2'] = { order = 302 },
                        ['storage-tank-3'] = { order = 303 },
                        ['storage-tank-4'] = { order = 304 },
                    }
                },
            }
        },
        ]]--
    }
end


mappings/base.lua
local belt_item_orders = {
    ['transport-belt'] = 100,
    ['underground-belt'] = 200,
    ['splitter'] = 300,
    ['loader'] = 1000,
}

local function scan_prototypes(result)
    -- Autogenerate quick-item-scroll groups.

    -- Currently, this only works for belts.
    local belt_groups = {}

    local function scan_prototype(name, item_prototype)
        -- Bail on hidden items.
        if item_prototype.flags and item_prototype.flags['hidden'] then return end
        local proto = item_prototype.place_result

        if not proto then return end
        if belt_groups then  -- Haven't already failed out.
            local speed, order  = proto.belt_speed, belt_item_orders[proto.type]
            if speed and order then
                --game.print(name .. ": speed=" .. speed .. "; order=" .. order)
                speed = speed * 480 -- Convert to items/sec
                if not belt_groups[speed] then
                    belt_groups[speed] = {
                        order = speed, items = {}, types = {}
                    }
                elseif belt_groups[speed].types[proto.type] then
                    belt_groups = nil  -- Failed autogen.
                end
                if belt_groups then
                    belt_groups[speed].items[name] = { order = order, type = proto.type }
                    belt_groups[speed].types[proto.type] = true
                end
            end
        end
    end

    for name, proto in pairs(game.item_prototypes) do
        scan_prototype(name, proto)
    end

    if belt_groups then
        for _, data in pairs(belt_groups) do data.types = nil end  -- Only used by us.
        --game.print(serpent.block(belt_groups))
        --log(serpent.block(belt_groups))
        result.categories.belts = { groups = belt_groups }
    else
        game.print("[quick-item-scroll] Failed to compute belt groups, falling back to defaults.  Belts from mods may not be supported.")
    end

    return result
end


return function()
    result = {
        categories = {
            belts = {
                groups = {
                    normal = {
                        order = 100,
                        items = {
                            ['transport-belt'   ] = { order = 100, type = 'belt' },
                            ['underground-belt' ] = { order = 200, type = 'underground' },
                            ['splitter'         ] = { order = 300, type = 'splitter' },
                        },
                    },
                    fast = {
                        order = 200,
                        items = {
                            ['fast-transport-belt'   ] = { order = 100, type = 'belt' },
                            ['fast-underground-belt' ] = { order = 200, type = 'underground' },
                            ['fast-splitter'         ] = { order = 300, type = 'splitter' },
                        },
                    },
                    express = {
                        order = 300,
                        items = {
                            ['express-transport-belt'   ] = { order = 100, type = 'belt' },
                            ['express-underground-belt' ] = { order = 200, type = 'underground' },
                            ['express-splitter'         ] = { order = 300, type = 'splitter' },
                        },
                    },
                }
            },
            fluids = {
                groups = {
                    pipes = {
                        order = 100,
                        items = {
                            ['pipe'] = { order = 100 },
                            ['pipe-to-ground'] = { order = 200 },
                        }
                    },
                    structures = {
                        order = 200,
                        items = {
                            ['offshore-pump'] = { order = 100 },
                            ['pump'] = { order = 200 },
                            ['storage-tank'] = { order = 300 },
                        }
                    },
                }
            },
            circuits = {
                groups = {
                    wires = {
                        order = 100,
                        items = {
                            ['green-wire'] = { order = 100 },
                            ['red-wire'] = { order = 200 },
                            ['copper-cable'] = { order = 300 },
                        }
                    },
                    combinators = {
                        order = 200,
                        items = {
                            ['constant-combinator'] = { order = 100 },
                            ['arithmetic-combinator'] = { order = 200 },
                            ['decider-combinator'] = { order = 300 },
                            ['small-lamp'] = { order = 1000 },
                            ['power-switch'] = { order = 1100 },
                            ['programmable-speaker'] = { order = 1200 },
                        }
                    },
                }
            },
            storage = {
                groups = {
                    normal = {
                        order = 100,
                        items = {
                            ['wooden-chest'] = { order = 100 },
                            ['iron-chest'] = { order = 200 },
                            ['steel-chest'] = { order = 300 },
                        }
                    },
                    logistics = {
                        order = 200,
                        items = {
                            ['logistic-chest-active-provider'] = { order = 100 },
                            ['logistic-chest-passive-provider'] = { order = 200 },
                            ['logistic-chest-storage'] = { order = 300 },
                            ['logistic-chest-buffer'] = { order = 400 },
                            ['logistic-chest-requester'] = { order = 500 },
                        }
                    },
                }
            },
            rail = {
                groups = {
                    track = {
                        order = 100,
                        items = {
                            ['rail'] = { order = 100 },
                            ['train-stop'] = { order = 200 },
                        }
                    },
                    signals = {
                        order = 200,
                        items = {
                            ['rail-signal'] = { order = 100 },
                            ['rail-chain-signal'] = { order = 200 },
                        }
                    },
                }
            },
            trains = {
                groups = {
                    trains = {
                        order = 100,
                        items = {
                            ['locomotive'] = { order = 100 },
                            ['cargo-wagon'] = { order = 200 },
                            ['fluid-wagon'] = { order = 300 },
                            ['artillery-wagon'] = { order = 400 },
                        }
                    },
                }
            },
            inserters = {
                groups = {
                    plain = {
                        order = 100,
                        items = {
                            ['burner-inserter'] = { order = 100 },
                            ['inserter'] = { order = 200 },
                            ['long-handed-inserter'] = { order = 300 },
                        }
                    },
                    fast = {
                        order = 200,
                        items = {
                            ['fast-inserter'] = { order = 100, type = 'normal' },
                            ['filter-inserter'] = { order = 200, type = 'filter' },
                        }
                    },
                    stack = {
                        order = 300,
                        items = {
                            ['stack-inserter'] = { order = 100, type = 'normal' },
                            ['stack-filter-inserter'] = { order = 200, type = 'filter' },
                        }
                    },
                }
            },
            power_production = {
                groups = {
                    steam = {
                        order = 100,
                        items = {
                            ['boiler'] = { order = 100 },
                            ['steam-engine'] = { order = 200 },
                        }
                    },
                    solar = {
                        order = 200,
                        items = {
                            ['solar-panel'] = { order = 100 },
                            ['accumulator'] = { order = 200 },
                        }
                    },
                    nuclear = {
                        order = 300,
                        items = {
                            ['nuclear-reactor'] = { order = 100 },
                            ['heat-exchanger'] = { order = 200 },
                            ['heat-pipe'] = { order = 300 },
                        }
                    },
                }
            },
            power_distribution = {
                groups = {
                    only_one_group = {
                        order = 100,
                        items = {
                            ['small-electric-pole'] = { order = 100 },
                            ['medium-electric-pole'] = { order = 200 },
                            ['big-electric-pole'] = { order = 300 },
                            ['substation'] = { order = 400 },
                        }
                    }
                }
            },
            miners = {
                groups = {
                    only_one_group = {
                        order = 100,
                        items = {
                            ['burner-mining-drill'] = { order = 100 },
                            ['electric-mining-drill'] = { order = 200 },
                        }
                    },
                }
            },
            assemblers = {
                groups = {
                    only_one_group = {
                        order = 100,
                        items = {
                            ['assembling-machine-1'] = { order = 100 },
                            ['assembling-machine-2'] = { order = 200 },
                            ['assembling-machine-3'] = { order = 300 },
                        }
                    },
                }
            },
            furnaces = {
                groups = {
                    only_one_group = {
                        order = 100,
                        items = {
                            ['stone-furnace'] = { order = 100 },
                            ['steel-furnace'] = { order = 200 },
                            ['electric-furnace'] = { order = 300 },
                        }
                    },
                }
            },
            tiles = {
                groups = {
                    only_one_group = {
                        order = 100,
                        items = {
                            ['stone-brick'] = { order = 100 },
                            ['concrete'] = { order = 200 },
                            ['hazard-concrete'] = { order = 300 },
                            ['landfill'] = { order = 400 },
                        }
                    },
                }
            },
            modules = {
                groups = {
                    tier1 = {
                        order = 100,
                        items = {
                            ['speed-module'] = { order = 100, type='speed' },
                            ['effectivity-module'] = { order = 200, type='effectivity'  },
                            ['productivity-module'] = { order = 300, type='productivity'  },
                        }
                    },
                    tier2 = {
                        order = 200,
                        items = {
                            ['speed-module-2'] = { order = 100, type='speed' },
                            ['effectivity-module-2'] = { order = 200, type='effectivity'  },
                            ['productivity-module-2'] = { order = 300, type='productivity'  },
                        }
                    },
                    tier3 = {
                        order = 300,
                        items = {
                            ['speed-module-3'] = { order = 100, type='speed' },
                            ['effectivity-module-3'] = { order = 200, type='effectivity'  },
                            ['productivity-module-3'] = { order = 300, type='productivity'  },
                        }
                    },
                }
            },
        },
    }

    return scan_prototypes(result)
end

